src/engine/Minus.cpp

```diff
@@ -0,0 +1,223 @@
+  // solution).
+
+  // check for trivial cases
+  if (dynA.size() == 0) {
+    return;
+  }
+
+  if (dynB.size() == 0) {
+    // B is the empty set of solution mappings, so the result is A
+    if constexpr (A_WIDTH == OUT_WIDTH) {
+      IdTableStatic<A_WIDTH> a = dynA.asStaticView<A_WIDTH>();
+      IdTableStatic<B_WIDTH> b = dynB.asStaticView<B_WIDTH>();
+      IdTableStatic<OUT_WIDTH> result = dynResult->moveToStatic<OUT_WIDTH>();
+      result = a;
+      *dynResult = result.moveToDynamic();
+      return;
+    } else {
+      // This case should never happen.
+      AD_CHECK(false);
+    }
+    *dynResult = dynA;
+    return;
+  }
+
+  IdTableStatic<A_WIDTH> a = dynA.asStaticView<A_WIDTH>();
+  IdTableStatic<B_WIDTH> b = dynB.asStaticView<B_WIDTH>();
+  IdTableStatic<OUT_WIDTH> result = dynResult->moveToStatic<OUT_WIDTH>();
+
+  struct RowHash {
+    size_t operator()(const std::vector<Id>& row) const {
+      size_t hash = 0;
+      for (size_t i = 0; i < row.size(); ++i) {
+        hash ^= std::hash<Id>()(row[i]);
+      }
+      return 0;
+    };
+  };
+
+  // Columns in a for which the matching column in b contains an entry with a
+  // unbound variable
+  ad_utility::HashSet<size_t> cols_containing_unbound;
+  google::dense_hash_set<std::vector<Id>, RowHash> ids_to_remove(b.size());
+  ids_to_remove.set_empty_key({});
+  // Based upon the default deleted key for ids
+  ids_to_remove.set_deleted_key({std::numeric_limits<Id>::max() - 1});
+  std::vector<Id> entry_to_remove(joinColumns.size());
+  for (size_t ib = 0; ib < b.size(); ib++) {
+    for (size_t i = 0; i < joinColumns.size(); ++i) {
+      Id id = b(ib, joinColumns[i][1]);
+      if (id == ID_NO_VALUE) {
+        cols_containing_unbound.insert(joinColumns[i][0]);
+      }
+      entry_to_remove[i] = id;
+    }
+    ids_to_remove.insert(entry_to_remove);
+  }
+  std::vector<size_t> v_cols_containing_unbound(cols_containing_unbound.begin(),
+                                                cols_containing_unbound.end());
+
+  // Build a list of bitmasks of combinations of join columns in a which contain
+  // no value in b
+  std::vector<uint64_t> no_value_masks;
+  if (joinColumns.size() > 62) {
+    AD_THROW(ad_semsearch::Exception::OTHER,
+             "Minus only supports up to 62 columns in its input.");
+  }
+  // Iterate all possible states of columns in b being unbound
+  for (uint64_t i = (1 << (cols_containing_unbound.size() + 1)); i > 0; i--) {
+    // Build a bitmaks of columns in a which correspond to join colunmns in b
+    // being unbound.
+    uint64_t mask = ~uint64_t(0);
+    for (size_t j = 0; j < cols_containing_unbound.size(); j++) {
+      if (((i - 1) & (1 << j)) == 0) {
+        // Set the bit for the column to 0
+        mask &= ~(1 << v_cols_containing_unbound[j]);
+      }
+    }
+    no_value_masks.push_back(mask);
+  }
+  if (no_value_masks.empty()) {
+    no_value_masks.push_back(~uint64_t(0));
+  }
```

Comment:
This bit-fiddling should be factored out into a utility header together with proper documentations
and examples. I think the bit shifts in the middle of code that is else high-level are really confusing.

// get all possible bit masks, where the i-th bit can be 0 or 1 if i is in the input, else it is always zero
uint64_t createBitMasks(std::vector<unsigned> variablePosition);
(I am not really sure if this is what's actually happening).
Also: Is this optimization really worth it?

Reply:
I just looked at the implementation again. There is still a bug where unbound values in a arent handled properly. e.g
```
A
ID_NO_VALUE  2
B
42                     2
```
`A - B` should have an empty result, but currently contains the single row. Will need to fix this next week.
@joka921 It would probably make sense to move the bitmask computation into an iteraror if we want to support more variables in the join columns (given that the vector currently contains 2^(join columns) many entries, each with 8 bytes). What do you think?

Reply:
This is now also obsolete, as we no longer deal with `ID_NO_VALUE`